rules/noUnnecessaryElseRule.ts

```diff
@@ -1,22 +1,54 @@
-import * as ts from 'typescript';
+// tslint
 import * as Lint from 'tslint';
-import * as utils from 'tsutils';
-import { AbstractIfStatementWalker } from '../src/walker';
+
+// tsrules.tslint-consistent-codestyle.src
 import { isElseIf } from '../src/utils';
 
+// tsutils
+import * as utils from 'tsutils';
+
+// typescript
+import * as ts from 'typescript';
+
 const FAIL_MESSAGE = `unnecessary else`;
+const FAIL_MESSAGE_BLOCK = `unnecessary else block`;
 
 export class Rule extends Lint.Rules.AbstractRule {
-    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
-        return this.applyWithWalker(new IfWalker(sourceFile, this.ruleName, undefined));
-    }
+  public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+    return this.applyWithWalker(new IfWalker(sourceFile, this.ruleName, undefined));
+  }
 }
 
-class IfWalker extends AbstractIfStatementWalker<void> {
-    protected _checkIfStatement(node: ts.IfStatement) {
-        if (node.elseStatement !== undefined &&
-            !isElseIf(node) &&
-            utils.endsControlFlow(node.thenStatement))
-            this.addFailureAtNode(node.getChildAt(5 /*else*/, this.sourceFile), FAIL_MESSAGE);
+class IfWalker extends Lint.AbstractWalker<void> {
+  public walk(sourceFile: ts.SourceFile) {
+    const cb = (node: ts.Node): void => {
+      if (node.kind === ts.SyntaxKind.IfStatement && !this._checkIfStatement(<ts.IfStatement>node))
+        return;
```

Comment:
returning here will not find errors inside `node.condition` and `node.thenStatement`
if the fixed range is smaller we could still inspect all children of this node

Reply:
Right, but this fixer may need multiple iterations in any case - if you have an ```else if``` cascade with returns in each block, ```endsControlFlow``` will only detect the first instance.

Reply:
You're right about the need for multiple iterations.
However, I think it's rather surprising if more failures keep popping up after fixing one. Instead I'd like to see all failures at once.

Also note that the following could be fixed in the same iteration if we didn't return early here:

```ts
if (a) {
  if (b) {
    return b;
  } else {
    return b;
  }
} else {
  return c;
}
```